include/aws/cal/rsa.h

```diff
@@ -0,0 +1,167 @@
+ * supported by the key - bytes needed for padding).
+ */
+AWS_CAL_API size_t aws_rsa_key_pair_max_encrypt_plaintext_size(
+    struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_encryption_algorithm algorithm);
+
+/*
+ * Uses the key_pair's private key to encrypt the plaintext. The output will be
+ * in out. out must be large enough to to hold the ciphertext. Check
+ * aws_rsa_key_pair_block_length() for output upper bound.
+ */
+AWS_CAL_API int aws_rsa_key_pair_encrypt(
+    struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_encryption_algorithm algorithm,
+    struct aws_byte_cursor plaintext,
+    struct aws_byte_buf *out);
+
+/*
+ * Uses the key_pair's private key to decrypt the ciphertext. The output will be
+ * in out. out must be large enough to to hold the ciphertext. Check
+ * aws_rsa_key_pair_block_length() for output upper bound.
+ */
+AWS_CAL_API int aws_rsa_key_pair_decrypt(
+    struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_encryption_algorithm algorithm,
+    struct aws_byte_cursor ciphertext,
+    struct aws_byte_buf *out);
+
+/*
+ * Max size for a block supported by a given key pair.
+ */
+AWS_CAL_API size_t aws_rsa_key_pair_block_length(struct aws_rsa_key_pair *key_pair);
+
+/**
+ * Uses the key_pair's private key to sign message. The output will be in out. out must be large enough
+ * to hold the signature. Check aws_rsa_key_pair_signature_length() for the appropriate size.
+ *
+ * It is the callers job to make sure message is the appropriate cryptographic digest for this operation. It's usually
+ * something like a SHA256.
+ */
+AWS_CAL_API int aws_rsa_key_pair_sign_message(
+    struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_signing_algorithm algorithm,
+    struct aws_byte_cursor digest,
+    struct aws_byte_buf *out);
+
+/**
+ * Uses the key_pair's public key to verify signature of message.
+ *
+ * It is the callers job to make sure message is the appropriate cryptographic digest for this operation. It's usually
+ * something like a SHA256.
+ *
+ * returns AWS_OP_SUCCESS if the signature is valid.
+ */
+AWS_CAL_API int aws_rsa_key_pair_verify_signature(
+    struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_signing_algorithm algorithm,
+    struct aws_byte_cursor digest,
+    struct aws_byte_cursor signature);
+
+/*
+ * Max size for a signature supported by a given key pair.
+ */
+AWS_CAL_API size_t aws_rsa_key_pair_signature_length(struct aws_rsa_key_pair *key_pair);
+
+enum aws_rsa_key_export_format { AWS_CAL_RSA_KEY_EXPORT_PKCS1 };
+
+/*
+ * Get public key for the key pair. Returns bytes of specified format.
+ * Any encoding on top of that (ex. b64) is left up to user.
+ * Note: this function is currently not supported on windows for generated keys.
+ */
+AWS_CAL_API int aws_rsa_key_pair_get_public_key(
+    const struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_key_export_format format,
+    struct aws_byte_cursor *out);
```

Comment:
This is going to get weird if we ever support another export format, since it's giving users the bytes by pointer.

If someone asked for the bytes in format A and then format B, where does that memory actually live?

Probably want to do `struct aws_byte_buf *out` and write the output to a user-owned buffer.

Or get rid of the idea of a customizable export format

Reply:
you are right. i started out with single format, but forgot to switch to buf after expanding. 
I prefer to keep the option to support other export formats in the future